core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSourceHandle.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.spi.exchange;
+
+/*
+ * Implementation is expected to be Jackson serializable and include equals, hashCode and toString methods
+ */
+public interface ExchangeSourceHandle
+{
+    int getPartitionId();
+
+    long getRetainedSizeInBytes();
+
+    @Override
+    boolean equals(Object obj);
+
+    @Override
+    int hashCode();
+
+    @Override
+    String toString();
```

Comment:
What's the purpose of these? Every object implements these methods, so they don't impose any requirement or constraint on implementations.

Reply:
Unfortunately it is impossible to enforce presence of these methods compile time. The idea behind leaving them explicitly declared in this interface is to make it more difficult to miss for somebody who is going to be implementing this interface. Though yeah, it doesn't provide any strong guarantees.